### PR TITLE
[Issue] Picking Handler Optimization Discussion

### PR DESCRIPTION
**Pull request by Strepto**
#### Type of change

This change is not for merging, but used as a discussion for two possible optimizations to increase performance when raycasting with multiple models.

See the comments on the code for more information. This "replaces" an issue, as I already toyed with the code here.

This issue is created as a follow up for a discussion with @christjt 

#### Jira ticket :blue_book:

N/A

## Description :pencil:

This PR adds one file out of a few that would be changed. Its just added for a discussion point and inspiration for a possible future change.

## How has this been tested? :mag:

N/A

## Test instructions :information_source:

Testing performance with multiple models, even with just a single model can highlight the performance differences between a hit and a non-hit on the bounding boxes.

## Checklist :ballot_box_with_check:

DO NOT MERGE THIS. Its not for merging

### CODE REVIEW

File: viewer/packages/cad-model/src/picking/PickingHandler.ts [hunk 1]
```diff
@@ -133,11 +189,9 @@ export class PickingHandler {
     if (treeIndex === undefined) {
       return undefined;
     }
-    const depth = await this.pickDepth(pickInput, async);
+    const perspectiveDepthNormalized = await this.pickDepth(pickInput, async);
```
Comment (Strepto) [inline viewer/packages/cad-model/src/picking/PickingHandler.ts:192]: Below here is refactorings only to reduce code duplication for the Depth only raycast

**Review by Strepto — COMMENTED** [review]
Comments on the two optimization approaches.

**Comment by christjt** [comment]
@Strepto Ok, so looking at this a little more high level, what I could tell here is this:

1.  Raycast the the boundingboxes of the models, if the ray doesn't intersect a bounding box then we don't need to do picking on it.
2. Sort the intersecting models bbs front to back and pick in a front-to-back order. If the picked distance is closer than the next bounding box(es) closest hit, then you can prune all behind models (i.e. no need to run picking).

Anything I missed here?

**Comment by Strepto** [comment]
> @Strepto Ok, so looking at this a little more high level, what I could tell here is this:
> 
> 1. Raycast the the boundingboxes of the models, if the ray doesn't intersect a bounding box then we don't need to do picking on it.

Yep.

> 2. Sort the intersecting models bbs front to back and pick in a front-to-back order. If the picked distance is closer than the next bounding box(es) closest hit, then you can prune all behind models (i.e. no need to run picking). (there is of course some edge-cases that need to be handled here as in the case where you are inside the bounding box of multiple models.)

If you are inside a bounding box the distance will be zero and the distance compared to the hit position will always be lower. But ofc needs validation. 

> Anything I missed here?

The depth only raycast avoids multiple re-renders as it does not care which model it hits, so it might be a useful addition.

Otherwise I think thats it.

**Comment by christjt** [comment]
> If you are inside a bounding box the distance will be zero and the distance compared to the hit position will always be lower. But ofc needs validation.

Right, but need to have some logic that guarantees any sectors that you are inside to be rendered. But should be quite trivial.

**Comment by pramodcog** [comment]
@Strepto and @christjt shall I close this PR as picking improvement discussed in here is implemented in https://github.com/cognitedata/reveal/pull/2970 & https://github.com/cognitedata/reveal/pull/2953 which are merged into the master?

**Comment by Strepto** [comment]
👍

**Comment by pramodcog** [comment]
Has been merged to master with https://github.com/cognitedata/reveal/pull/2970 and https://github.com/cognitedata/reveal/pull/2953

**Pull request closed without merging**